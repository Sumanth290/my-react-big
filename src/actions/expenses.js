import uuid from "uuid";

export const addExpense = ( {
    desc = "",
    note = "",
    amount = 0,
    timestamp = 0
} = {
    desc : "",
    note : "",
    amount : 0,
    timestamp : 0
} ) => ({
    type : "ADD_EXPENSE",
    expense : {
        id : uuid(),
        desc,
        note,
        amount,
        timestamp
    }
});


export const removeExpense = ( id ) => ({
    type : "REMOVE_EXPENSE",
    id
});


export const editExpense = (id,updates = {}) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});