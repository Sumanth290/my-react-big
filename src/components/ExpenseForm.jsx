import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            desc : props.expense ? props.expense.desc : "",
            note : props.expense ? props.expense.note : "",
            amount : props.expense ? (props.expense.amount / 100).toString() :"",
            timestamp : props.expense ? moment(props.expense.timestamp) : moment(),
            calendarFocused : false,
            error : ""
        }
    }

    handleDescOnChange = (e) => {
        const desc = e.target.value;
        this.setState(() => ({ desc }));
    };

    handleNoteOnChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    handleAmtOnChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }
    };

    handleDateChange = (timestamp) => {
        if(timestamp){
            this.setState(() => ({ timestamp }));
        }
    };

    handleFocusChange = ({focused}) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    handleAddExpenseButtonClick = (e) => {
        if(!this.state.desc || !this.state.amount)
        {
            this.setState(() => ({error : "Please provide 'Description' and 'Amount'"}));
        }
        else{
            this.setState(() => ({error : ""}));
            this.props.onSubmit({
                desc : this.state.desc,
                note : this.state.note,
                amount : parseFloat(this.state.amount,10) * 100 ,
                timestamp : this.state.timestamp.valueOf()
            });
            console.log("Submitted !!");
        }
    }

    render(){
        return (
            <div className="form">
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input 
                    className="text-input"
                    type="text" 
                    placeholder="Description" 
                    autoFocus 
                    value={this.state.desc}
                    onChange={this.handleDescOnChange}
                />
                <input 
                    className="text-input"
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.handleAmtOnChange}
                />
                <SingleDatePicker
                    date={this.state.timestamp} 
                    onDateChange={this.handleDateChange} 
                    focused={this.state.calendarFocused}
                    onFocusChange={this.handleFocusChange}
                    isOutsideRange={() => ( false )}
                    numberOfMonths={1}
                />
                <textarea
                    className="text-area"
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.handleNoteOnChange}
                ></textarea>
                <div>
                    <button className="buttons form_button" onClick={this.handleAddExpenseButtonClick}>Save Expense</button>
                </div>
            </div>
        );
    }
}