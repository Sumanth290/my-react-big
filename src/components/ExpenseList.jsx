import React from "react";
import { connect } from "react-redux";
import getVisibleExpenses from "../selectors/expenses";
import ExpenseListItem from "./ExpenseListItem.jsx";

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses && props.expenses.length !== 0 ? (
                <ul>
                    {props.expenses.map((expense) => (
                        <li key={expense.id}><ExpenseListItem {...expense} /></li>
                    ))}
                </ul>
            ) : (
                <p>No Expenses</p>
            )
        }
    </div>
);

const mapStateToProps = (state) => ({
    expenses : getVisibleExpenses(state.expenses,state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
