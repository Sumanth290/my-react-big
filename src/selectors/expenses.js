
export default (expenses,{text,sortBy,startDate,endDate}) => {
    return expenses.filter((expense) => {
        const startDateMatch = startDate ? startDate.isSameOrBefore(expense.timestamp,"day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(expense.timestamp,"day") : true;
        const textMatch = typeof text !== "string" || (
            (!!expense.desc && expense.desc.toLowerCase().includes(text.toLowerCase())) ||
            (!!expense.note && expense.note.toLowerCase().includes(text.toLowerCase()))
        ) ;
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy === "amount"){
            return b.amount - a.amount;
        }
        else{
            return a.timestamp > b.timestamp ? -1 : 1 ;
        }
    });
};