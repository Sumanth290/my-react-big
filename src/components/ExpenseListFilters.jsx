import React from "react";
import { connect } from "react-redux";
import {setTextFilter,sortByAmount,sortBydate,setStartDate,setEndDate} from "../actions/filters";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component{
    constructor(props) {
        super(props);
        this.state = {
            startDate: null,
            endDate: null,
            calendarFocused: null,
        };
    }
    
    handleSortFilterChange = (e) => {
        (e.target.value && e.target.value === "amount") ? 
            this.props.sortByAmountDispatcher() : 
            this.props.sortBydateDispatcher()
    };

    handleDateChange = ({startDate,endDate}) => {
        this.setState(() => ({ startDate,endDate }));
        this.props.startDateDispatcher(startDate);
        this.props.endDateDispatcher(endDate);
    };

    handleFocusChange = (focused) => {
        console.log(focused);
        this.setState(() => ({ calendarFocused: focused}));
    };

    handleTextChange = (e) => {
        this.props.textFilterDispatcher(e.target.value);
    };
    
    render(){
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            type="text" 
                            value={this.props.filters.text} 
                            onChange={this.handleTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <div>
                            <input 
                                type="radio" 
                                name="sortFilter" 
                                onChange={this.handleSortFilterChange} 
                                value="date" 
                                checked={this.props.filters.sortBy === "date"}
                            />Date
                        </div>
                        <div>
                            <input 
                                type="radio" 
                                name="sortFilter" 
                                onChange={this.handleSortFilterChange} 
                                value="amount" 
                                checked={this.props.filters.sortBy === "amount"}
                            />Amount
                        </div>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate} 
                            startDateId="filterStartDate"
                            endDate={this.props.filters.endDate} 
                            endDateId="filterEndDate"
                            onDatesChange={this.handleDateChange} 
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.handleFocusChange}
                            showClearDates={true}
                            isOutsideRange={() => ( false )}
                            numberOfMonths={1}
                        />
                    </div>
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => ({filters : state.filters});

const mapDispatchToProps = (dispatch) => ({
    sortByAmountDispatcher : () => dispatch(sortByAmount()),
    sortBydateDispatcher : () => dispatch(sortBydate()),
    startDateDispatcher : (startDate) => dispatch(setStartDate(startDate)),
    endDateDispatcher : (endDate) => dispatch(setEndDate(endDate)),
    textFilterDispatcher : (text) => dispatch(setTextFilter(text))
});

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);