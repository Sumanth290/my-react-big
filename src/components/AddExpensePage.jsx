import React from "react";
import ExpenseForm from "./ExpenseForm.jsx";
import { connect } from "react-redux";
import {addExpense} from "../actions/expenses";

export class AddExpensePage extends React.Component{
    onSubmit = (expense) => {
        this.props.addExpenseDispatcher(expense);
        this.props.history.push("/");
    };

    render(){
        return (
            <div>
                <h2>This is "Add Expense Page"</h2>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpenseDispatcher : (expense) => dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage);