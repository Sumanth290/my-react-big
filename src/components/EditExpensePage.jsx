import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm.jsx";
import {startEditExpense,startRemoveExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component{
    handleOnSubmit = (expense) => {
        this.props.editExpenseDispatcher(this.props.expense.id,expense);
        this.props.history.push("/dashboard");
    };

    handleRemove = (e) => {
        if(confirm("Are you sure you want to remove expense ?")){
            this.props.removeExpenseDispatcher(this.props.expense.id);
            console.log("Removed !!");
            this.props.history.push("/dashboard");
        }
    };

    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <div className="page-header__content">
                            <p className="page-header__title">Edit Expense</p>
                        </div>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={this.handleOnSubmit}
                    />
                    <button className="buttons buttons--remove" onClick={this.handleRemove}>Remove Expense</button>
                </div>
            </div>
        );
    }
}


const mapStatetoProps = (state,props) => ({
    expense : state.expenses.find((expense) => (expense.id === props.match.params.id))
});

const mapDispatchToProps = (dispatch) => ({
    editExpenseDispatcher : (id,expense) => dispatch(startEditExpense(id,expense)),
    removeExpenseDispatcher : (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStatetoProps,mapDispatchToProps)(EditExpensePage);