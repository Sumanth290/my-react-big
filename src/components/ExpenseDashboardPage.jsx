import React from "react";
import ExpenseList from "./ExpenseList.jsx";
import ExpenseListFilters from "./ExpenseListFilters.jsx";


const ExpenseDashboardPage = () => (
    <div>
        <h2>This is "Expense Dashboard Home Page"</h2>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;