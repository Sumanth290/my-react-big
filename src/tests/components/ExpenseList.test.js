import React from "react";
import {shallow} from "enzyme";
import {ExpenseList} from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("Should render expenseList with expenses",()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test("Should render no expenses without expenses props",()=>{
    const wrapper = shallow(<ExpenseList />);
    expect(wrapper).toMatchSnapshot();
});

test("Should render no expenses with empty expenses props",()=>{
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});