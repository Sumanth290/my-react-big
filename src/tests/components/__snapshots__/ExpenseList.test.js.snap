// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expenseList with expenses 1`] = `
<div>
  <ul>
    <li
      key="1"
    >
      <ExpenseListItem
        amount={134}
        desc="gum"
        id="1"
        note=""
        timestamp={"1970-01-01T00:00:00.000Z"}
      />
    </li>
    <li
      key="2"
    >
      <ExpenseListItem
        amount={109500}
        desc="rent"
        id="2"
        note=""
        timestamp={"1969-12-28T00:00:00.000Z"}
      />
    </li>
    <li
      key="3"
    >
      <ExpenseListItem
        amount={4500}
        desc="Credit Card"
        id="3"
        note=""
        timestamp={"1970-01-05T00:00:00.000Z"}
      />
    </li>
  </ul>
</div>
`;

exports[`Should render no expenses with empty expenses props 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`Should render no expenses without expenses props 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
